File: windows/DoNotType.Core.Tests/LiveModeTests.cs
using DoNotType.Core;
using Xunit;

namespace DoNotType.Core.Tests;

/// <summary>
/// The three modes a dictation can be started in, asserted in the same shape as
/// <c>Tests/DoNotTypeCoreTests/LiveModeTests.swift</c> and
/// <c>android/app/src/test/kotlin/app/donottype/core/LiveModeTest.kt</c>.
/// </summary>
/// <remarks>
/// The table is duplicated rather than shared on purpose: a fixture read from disk would be read
/// by whichever platform remembered to read it, and this is the file that says a phone and a
/// laptop are the same product -- see docs/PARITY.md.
///
/// On this client the picker is three hot keys rather than a chip, which is exactly why the rule
/// had to move here: Windows resolved the stage with a conditional of its own that read the target
/// language on every dictation, so setting one took the main key away from verbatim.
/// </remarks>
public sealed class LiveModeTests
{
    private static readonly Func<ProviderKind, bool> Unkeyed = _ => false;
    private static readonly Func<ProviderKind, bool> Keyed = _ => true;

    /// <summary>The persisted spelling is what a log line and a settings document carry.</summary>
    [Fact]
    public void TheSpellingsAreStableAndUnknownValuesFallBack()
    {
        Assert.Equal(
            new[] { "dictate", "rewrite", "translate" },
            Enum.GetValues<LiveMode>().Select(mode => mode.Id()));
        foreach (var mode in Enum.GetValues<LiveMode>())
        {
            Assert.Equal(mode, LiveModeExtensions.From(mode.Id()));
        }
        Assert.Equal(LiveModeExtensions.Default, LiveModeExtensions.From("nonsense"));
        Assert.Equal(LiveModeExtensions.Default, LiveModeExtensions.From(null));
        Assert.Equal(LiveMode.Dictate, LiveModeExtensions.Default);
    }

    /// <summary>Dictation is the product; a default anywhere else changes a fresh install.</summary>
    [Fact]
    public void TheDefaultIsAPlainDictationWhateverElseIsConfigured()
    {
        Assert.Equal(
            TranscriptMode.Verbatim,
            LiveModeExtensions.Default.Stage(RewriteStyle.Formal, "French"));
    }

    [Fact]
    public void EachModeAsksForItsOwnStage()
    {
        Assert.Equal(
            TranscriptMode.Verbatim, LiveMode.Dictate.Stage(RewriteStyle.Formal, "French"));
        Assert.Equal(
            TranscriptMode.Rewrite(RewriteStyle.Formal),
            LiveMode.Rewrite.Stage(RewriteStyle.Formal, "French"));
        Assert.Equal(
            TranscriptMode.Translate("French"),
            LiveMode.Translate.Stage(RewriteStyle.Formal, "French"));
    }

    /// <summary>
    /// The exclusivity the three keys exist to make visible. A target language used to override
    /// both of the other keys from Settings, so the main key delivered a translation.
    /// </summary>
    [Fact]
    public void TranslateAndRewriteCannotHappenAtOnce()
    {
        foreach (var mode in Enum.GetValues<LiveMode>())
        {
            var stage = mode.Stage(RewriteStyle.Formal, "French");
            Assert.False(
                stage is TranscriptMode.RewriteMode && stage is TranscriptMode.TranslateMode,
                $"{mode} asked for two second stages");
        }
    }

    /// <summary>A mode with nothing configured is a dictation, not an unspecified request.</summary>
    [Fact]
    public void AnUnconfiguredModeFallsBackToTheTranscript()
    {
        Assert.Equal(TranscriptMode.Verbatim, LiveMode.Translate.Stage(RewriteStyle.Formal, ""));
        Assert.Equal(TranscriptMode.Verbatim, LiveMode.Translate.Stage(RewriteStyle.Formal, "   "));
        Assert.Equal(
            TranscriptMode.Verbatim, LiveMode.Rewrite.Stage(RewriteStyle.Verbatim, "French"));
    }

    /// <summary>Dictation has no second stage, so nothing here can be missing.</summary>
    [Fact]
    public void APlainDictationIsAlwaysAvailable()
    {
        var availability = LiveMode.Dictate.Availability(ProviderKind.Gemini, "", Unkeyed);
        Assert.True(availability.IsAvailable);
        Assert.Empty(availability.Reason);
    }

    /// <summary>The two backend-shaped answers are worded for the job that was chosen.</summary>
    [Fact]
    public void TheReasonNamesTheJobTheUserAskedFor()
    {
        Assert.Equal(
            "Add an API key first — without one nothing can run, rewriting included.",
            LiveMode.Rewrite.Availability(ProviderKind.Gemini, "", Unkeyed).Reason);
        Assert.Equal(
            "Add an API key first — without one nothing can run, translating included.",
            LiveMode.Translate.Availability(ProviderKind.Gemini, "French", Unkeyed).Reason);
        Assert.Equal(
            "Deepgram only transcribes audio and cannot rewrite text. Add a key for a backend "
            + "that can, and rewriting will use it.",
            new RewriteAvailability.BackendCannotRewrite(
                ProviderKind.Deepgram, SecondStageJob.Rewriting).Reason);
        Assert.Equal(
            "Deepgram only transcribes audio and cannot translate text. Add a key for a backend "
            + "that can, and translating will use it.",
            new RewriteAvailability.BackendCannotRewrite(
                ProviderKind.Deepgram, SecondStageJob.Translating).Reason);
    }

    /// <summary>
    /// Translate with nothing to translate into: the one state the old arrangement could not
    /// represent, because a target language was the switch rather than the destination.
    /// </summary>
    [Fact]
    public void TranslateWithoutALanguageIsUnavailableForThatReasonAlone()
    {
        Assert.Equal(
            new RewriteAvailability.NoTargetLanguage(),
            LiveMode.Translate.Availability(ProviderKind.Gemini, "  ", Keyed));
        Assert.Equal(
            "Set a target language in Settings first, and Translate will write in it.",
            new RewriteAvailability.NoTargetLanguage().Reason);
        Assert.True(
            LiveMode.Translate.Availability(ProviderKind.Gemini, "French", Keyed).IsAvailable);
    }

    /// <summary>68dp on Android, 86pt on iOS. Both bars are laid out for these three words.</summary>
    [Fact]
    public void TheLabelsAreShortEnoughForTheChip()
    {
        foreach (var mode in Enum.GetValues<LiveMode>())
        {
            Assert.NotEmpty(mode.Label());
            Assert.True(mode.Label().Length <= 9, mode.Label());
        }
    }
}

File: Tests/DoNotTypeCoreTests/LiveModeTests.swift
import XCTest

@testable import DoNotTypeCore

/// The phone keyboards' mode picker.
///
/// The same cases are asserted in `android/app/src/test/kotlin/app/donottype/core/LiveModeTest.kt`
/// and `windows/DoNotType.Core.Tests/CoreTests.cs` — see `docs/PARITY.md`. Windows has no picker,
/// but it ships the settings transfer that carries the value between the phones, so it has to
/// agree about the spellings.
final class LiveModeTests: XCTestCase {

    /// The persisted spelling is what the settings transfer writes, so a rename here silently
    /// resets every phone that reads a file written by an older build.
    func testTheSpellingsAreStableAndUnknownValuesFallBack() {
        XCTAssertEqual(LiveMode.allCases.map(\.rawValue), ["dictate", "rewrite", "translate"])
        XCTAssertEqual(LiveMode(rawValue: "nonsense"), nil)
        XCTAssertEqual(LiveMode.default, .dictate)
    }

    /// Dictation is the product. A mode picker that starts anywhere else changes what the app does
    /// on a fresh install.
    func testTheDefaultIsAPlainDictation() {
        XCTAssertEqual(
            LiveMode.default.stage(style: .formal, language: "French"), .verbatim,
            "neither a configured style nor a configured language may start a second stage")
    }

    func testEachModeAsksForItsOwnStage() {
        XCTAssertEqual(LiveMode.dictate.stage(style: .formal, language: "French"), .verbatim)
        XCTAssertEqual(LiveMode.rewrite.stage(style: .formal, language: "French"), .rewrite(.formal))
        XCTAssertEqual(
            LiveMode.translate.stage(style: .formal, language: "French"), .translate("French"))
    }

    /// The exclusivity the picker exists to make visible. A target language used to override the
    /// rewrite toggle from Settings, so the chip said Rewrite over a request that translated.
    func testTranslateAndRewriteCannotHappenAtOnce() {
        for mode in LiveMode.allCases {
            let stage = mode.stage(style: .formal, language: "French")
            let isRewrite = stage == .rewrite(.formal)
            let isTranslate = stage == .translate("French")
            XCTAssertFalse(isRewrite && isTranslate, "\(mode) asked for two second stages")
        }
    }

    /// An unconfigured mode is a dictation rather than a request that asks a model to do something
    /// unspecified to a transcript. The picker greys these out — `availability` is what it asks —
    /// but a stale persisted value must not turn into a strange request either.
    func testAnUnconfiguredModeFallsBackToTheTranscript() {
        XCTAssertEqual(LiveMode.translate.stage(style: .formal, language: ""), .verbatim)
        XCTAssertEqual(LiveMode.translate.stage(style: .formal, language: "   "), .verbatim)
        XCTAssertEqual(LiveMode.rewrite.stage(style: .verbatim, language: "French"), .verbatim)
    }

    /// 68dp on Android, 86pt on iOS. Both are laid out for these three words.
    func testTheLabelsAreShortEnoughForTheChip() {
        for mode in LiveMode.allCases {
            XCTAssertFalse(mode.label.isEmpty)
            XCTAssertLessThanOrEqual(mode.label.count, 9, mode.label)
        }
    }
}
